Add practical lessons from MMseqs2 tool wrapper development (#9)
galaxyproject/galaxy-skills: tool-dev/SKILL.md

**Parameter parity rule:** Every param in `<inputs>` must appear in `<command>`, and every flag in `<command>` must trace back to an `<inputs>` param or a macro token. Orphaned params are a lint warning and a review flag.

### Galaxy `.dat` Extension and Format Auto-Detection

> Galaxy stores datasets internally with a `.dat` extension (e.g., `dataset_d562...7f33.dat`). Tools that auto-detect input format from the file extension (like MMseqs2) will fail because `.dat` is not a recognized bioinformatics format. The fix is to create symlinks with the correct extension derived from Galaxy's datatype:

Reviewer: ```suggestion
Galaxy stores datasets internally with a `.dat` extension (e.g., `dataset_d562...7f33.dat`). Tools that auto-detect input format from the file extension will fail because `.dat` is not a recognized bioinformatics format. The fix is to create symlinks with the correct extension derived from Galaxy's datatype:
```

Author: Applied (reworded intro) as part of the `is_of_type()` rewrite (1da1c5c).

```xml
<command detect_errors="aggressive"><![CDATA[

> #set $query_ext = str($query_fasta.ext).replace('sanger', '').replace('illumina', '')

Reviewer: Puh, this is wild. If we know is sanger or illumina, we can simply call it `fastq`.

Author: Agreed — dropped the `.ext.replace()` hack; now branches on `is_of_type()` (1da1c5c).

> ln -s '$query_fasta' 'query.${query_ext}' &&

Author: Applied this suggestion (1da1c5c).

mytool search 'query.${query_ext}' ...
]]></command>
```

The `str($var.ext).replace('sanger', '').replace('illumina', '')` pattern maps Galaxy datatypes to standard file extensions:

| Galaxy datatype | `$var.ext` | After replace | Symlink |
|----------------|-----------|---------------|---------|
| fasta | `fasta` | `fasta` | `query.fasta` |
| fasta.gz | `fasta.gz` | `fasta.gz` | `query.fasta.gz` |
| fastqsanger | `fastqsanger` | `fastq` | `query.fastq` |
| fastqsanger.gz | `fastqsanger.gz` | `fastq.gz` | `query.fastq.gz` |
| fastqillumina | `fastqillumina` | `fastq` | `query.fastq` |

> This is especially common for tools that accept both FASTA and FASTQ, both compressed and uncompressed. Always symlink with the dynamic extension — hardcoding `.fasta` breaks when users provide gzipped or FASTQ inputs.

Reviewer: please remove that, way to complicated imho

Author: Removed that paragraph and the datatype→extension table (1da1c5c).

Use `format="markdown"` for new tools (preferred over RST). Structure with bold `**headers**`, horizontal rules `-----` between sections, double backticks for code references, and end with a citation block. Keep it concise and actionable.

> **Lint caveat:** Even with `format="markdown"`, `planemo lint` still validates help content as RST. RST grid tables with alignment issues will fail the `HelpValidRST` check. Use simple bullet lists or definition lists instead of complex tables in help sections.

Reviewer: Can you report this please as plenmo issue.

Author: Filed as galaxyproject/planemo#1650 and linked it next to the caveat (commit fe03f42).

### Named Yields and Token Parameterization

> #### Unnamed `<yield/>`

Author: Removed the "Unnamed `<yield/>`" subsection. Added a one-line note preferring a small macro expanded twice over a shared unnamed yield (1da1c5c).

<param name="reads" type="data" format="fastqsanger,fastqsanger.gz,fastqsanger.bz2" label="Reads"/>
```

> **FASTQ format naming:** Always use `fastqsanger` or `fastqillumina` instead of generic `fastq`. Galaxy distinguishes FASTQ quality encoding variants and reviewers will flag the generic name. For tools accepting any FASTQ encoding, list both: `format="fastqsanger,fastqsanger.gz,fastqillumina,fastqillumina.gz"`.

Reviewer: ```suggestion
**FASTQ format naming:** Always use `fastqsanger` or `fastqillumina` instead of generic `fastq`. Galaxy distinguishes FASTQ quality encoding variants and reviewers will flag the generic name. For tools accepting compressed FASTQ and both encoding, list all 4 variants: `format="fastqsanger,fastqsanger.gz,fastqillumina,fastqillumina.gz"`.
```

Author: Applied — this 4-variant guidance is in the current revision.

<param argument="--memory" type="integer" value="4" label="Memory (GB)"/>
```

> ### Out-of-Memory Detection

Reviewer: I would remove that.

the regex must match real text on the stderr, adding this when a tool does not print this text, will not help.

Author: Removed the OOM `<stdio>` regex section entirely (1da1c5c). Agree canned regexes that don't match a tool's actual stderr don't help.

Run `planemo test --biocontainers` to execute tests. See **Reference: Useful Planemo Commands** at the end of this document for the full command set.

**Fallback when Docker is unavailable:** If `--biocontainers` fails with exit code 127 (Docker not found), install the tool via conda and run without dependency resolution:

```bash

> conda create -p /path/to/env -c conda-forge -c bioconda <tool>=<version>

Reviewer: ```suggestion
```

Author: Removed the `conda create` line (1da1c5c).

> PATH="/path/to/env/bin:$PATH" planemo test --no_conda_auto_init --no_dependency_resolution tools/mytool/

Reviewer: planemo should be able to create the environment more precisely 

```suggestion
PATH="/path/to/env/bin:$PATH" planemo test tools/mytool/
```

Author: Simplified the fallback to `PATH="/path/to/env/bin:$PATH" planemo test tools/mytool/` (1da1c5c).
